lighthouse-core/lib/tracehouse/trace-processor.js

```diff
@@ -66,10 +66,68 @@ class TraceProcessor {
   }
 
   /**
+   * This method sorts a group of trace events that have the same timestamp. We want to...
+   *
+   * 1. Put E events first, we finish off our existing events before we start new ones.
+   * 2. Order B/X events by their duration, we want parents to start before child events.
+   * 3. If we don't have any of this to go on, just use the position in the original array (stable sort).
+   *
+   * @param {number[]} tsGroupIndices
+   * @param {number[]} timestampSortedIndices
+   * @param {number} indexOfTsGroupIndicesStart
+   * @param {LH.TraceEvent[]} traceEvents
+   * @return {number[]}
+   */
+  static _sortTimestampEventGroup(
+    tsGroupIndices,
+    timestampSortedIndices,
+    indexOfTsGroupIndicesStart,
+    traceEvents
+  ) {
+    const eEventIndices = tsGroupIndices.filter(index => traceEvents[index].ph === 'E');
+    const bxEventIndices = tsGroupIndices.filter(index => traceEvents[index].ph === 'X' ||
+      traceEvents[index].ph === 'B');
+    const otherEventIndices = tsGroupIndices.filter(index => !eEventIndices.includes(index) &&
+      !bxEventIndices.includes(index));
+
+    const effectiveDuration = new Map();
+    for (const index of bxEventIndices) {
+      const event = traceEvents[index];
+      if (event.ph === 'X') {
+        effectiveDuration.set(index, event.dur);
+      } else {
+        // Find the first 'E' event that matches our name.
+        const startIndex = indexOfTsGroupIndicesStart + tsGroupIndices.length;
+        for (let j = startIndex; j < timestampSortedIndices.length; j++) {
+          const potentialEndEvent = traceEvents[timestampSortedIndices[j]];
+          if (potentialEndEvent.ph === 'E' && potentialEndEvent.name === event.name) {
+            effectiveDuration.set(index, potentialEndEvent.ts - event.ts);
+            break;
+          }
+        }
+
+        if (!effectiveDuration.has(index)) {
```

Comment:
could avoid this `has()` awkwardness by setting a local `let duration = Number.MAX_SAFE_INTEGER`, updating it in the loop, then always setting it here :)